src/object/srv_obj.c

```diff
@@ -1003,7 +1003,18 @@ obj_singv_ec_rw_filter(daos_unit_oid_t oid, struct daos_oclass_attr *oca,
 				reentry = true;
 			} else {
 				D_ASSERT(!reentry);
-				iod->iod_recxs = (void *)iod->iod_size;
+				/* VOS_EC_SINGV_UPDATE_SKIP for the case for example update 2 akeys
+				 * singv in one IO, first long singv distributed to all shards,
+				 * second short singv only stored on one data shard and all parity
+				 * shard, should skip the second update on some data shards.
+				 */
+				if (nr > 1 && tgt_off != OBJ_EC_SHORT_SINGV_IDX &&
+				    is_ec_data_shard_by_tgt_off(tgt_off, oca) &&
+				    iod->iod_size <=
+				    OBJ_EC_SINGV_EVENDIST_SZ(obj_ec_data_tgt_nr(oca)))
+					iod->iod_recxs = (void *)VOS_EC_SINGV_UPDATE_SKIP;
```

Comment:
Can we set iod_recxs to NULL to skip this iod? instead of a new flag?

Btw: I thought obj_get_iods_offs_by_oid() should skip this IOD anyway? or there is a bug there? Maybe fix there?

Reply:
"Can we set iod_recxs to NULL to skip this iod?"
normal replica singv fetch with NULL iod_recxs, so cannot use NULL as an indicator.

"I thought obj_get_iods_offs_by_oid() should skip this IOD anyway? or there is a bug there? Maybe fix there?"
yes, I tested to skip iod/akey in obj_get_iods_offs_by_oid(), but a following problem is when doing bulk transferring, the client-side bulk handles array (one bulk handle for each iod/akey) was still same as original order, so we need another data structure to store the info of which set of iod/akey skipped and then skip corresponding client bulk handles when doing bulk transferring, that need some extra trivial code and seems worse than this patch's method.